### PR TITLE
Add support for IotHub resource

### LINKED ISSUE #1292: Feature: Will you add Azure IoT Hub support?
I would like to know if Azure IoT Hub is in the roadmap to be added as a resource.

### PR DESCRIPTION
**Pull request by super-harsh**
Closes #1292 

**What this PR does / why we need it**:

This PR adds support for `Devices/IotHub` 

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/api/devices/customizations/iot_hub_extension.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/iothub/armiothub"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	storage "github.com/Azure/azure-service-operator/v2/api/devices/v1api20210702storage"
```
Comment (matthchr) [inline v2/api/devices/customizations/iot_hub_extension.go:19]: minor: call this `iothub` or something? Calling it storage makes me think Azure storage.

File: v2/api/devices/customizations/iot_hub_extension.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+func secretsToWrite(obj *storage.IotHub, keys map[string]armiothub.SharedAccessSignatureAuthorizationRule) ([]*v1.Secret, error) {
+	operatorSpecSecrets := obj.Spec.OperatorSpec.Secrets
+	if operatorSpecSecrets == nil {
+		return nil, errors.Errorf("unexpected nil operatorspec")
+	}
+
+	collector := secrets.NewCollector(obj.Namespace)
+	iothubowner, ok := keys["iothubowner"]
```
Comment (matthchr) [inline v2/api/devices/customizations/iot_hub_extension.go:131]: Where did you get these values? Is it in some IOTHub documentation? Can you link it here? Or did you just get it by observing the payload of the response?

File: v2/api/devices/customizations/iot_hub_extension.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	if ok {
+		collector.AddValue(operatorSpecSecrets.IotHubOwnerPrimaryKey, *iothubowner.PrimaryKey)
```
Comment (matthchr) [inline v2/api/devices/customizations/iot_hub_extension.go:133]: Use `to.Value` on all of these, not direct dereference, or else there's a risk of panic if the field isn't set and we try to dereference it.

`collector.AddValue` already discards empty string values so it should just work then.

File: v2/api/devices/v1api20210702/structure.txt [hunk 4]
```diff
@@ -0,0 +1,841 @@
+// Code generated by azure-service-operator-codegen. DO NOT EDIT.
+github.com/Azure/azure-service-operator/v2/api/devices/v1api20210702
+├── APIVersion: Enum (1 value)
+│   └── "2021-07-02"
+├── IotHub: Resource
+│   ├── Owner: github.com/Azure/azure-service-operator/v2/api/resources/v1apiv20191001.ResourceGroup
+│   ├── Spec: Object (9 properties)
+│   │   ├── AzureName: string
+│   │   ├── Etag: *string
```
Comment (matthchr) [inline v2/api/devices/v1api20210702/structure.txt:9]: minor: Mark this field as readonly in azure-arm?

File: v2/azure-arm.yaml [hunk 5]
```diff
@@ -984,6 +984,34 @@ objectModelConfiguration:
           $armReference: true
         PrivateDnsZoneArmResourceId:
           $armReference: true
+  devices:
+    2021-07-02:
+      IotHub:
```
Comment (matthchr) [inline v2/azure-arm.yaml:989]: Should we also support EventHubConsumerGroups? That seems to be the other main IOTHub resource I can see in their Swagger.

File: v2/api/devices/v1api20210702/structure.txt [hunk 6]
```diff
@@ -0,0 +1,841 @@
...
+│   │   │       ├── RegistryReadWritePrimaryKey: *genruntime.SecretDestination
+│   │   │       ├── RegistryReadWriteSecondaryKey: *genruntime.SecretDestination
+│   │   │       ├── ServicePrimaryKey: *genruntime.SecretDestination
+│   │   │       └── ServiceSecondaryKey: *genruntime.SecretDestination
+│   │   ├── Owner: *genruntime.KnownResourceReference
+│   │   ├── Properties: *Object (19 properties)
+│   │   │   ├── AllowedFqdnList: string[]
+│   │   │   ├── AuthorizationPolicies: Object (4 properties)[]
+│   │   │   │   ├── KeyName: *string
+│   │   │   │   ├── PrimaryKey: *string
```
Comment (matthchr) [inline v2/api/devices/v1api20210702/structure.txt:36]: Is this used to configure the primary key of the policy or something? Or is it readonly?

Feels like either this whole `AuthorizationPolicies` thing should be marked readonly (if it is?), or else the Primary/Secondary key fields should be marked with `x-ms-secret`

File: v2/api/devices/v1api20210702/structure.txt [hunk 7]
```diff
@@ -0,0 +1,841 @@
...
+│   │   │   │       │   └── "Allow"
+│   │   │   │       ├── FilterName: *string
+│   │   │   │       └── IpMask: *string
+│   │   │   ├── PublicNetworkAccess: *Enum (2 values)
+│   │   │   │   ├── "Disabled"
+│   │   │   │   └── "Enabled"
+│   │   │   ├── RestrictOutboundNetworkAccess: *bool
+│   │   │   ├── Routing: *Object (4 properties)
+│   │   │   │   ├── Endpoints: *Object (4 properties)
+│   │   │   │   │   ├── EventHubs: Object (9 properties)[]
```
Comment (matthchr) [inline v2/api/devices/v1api20210702/structure.txt:106]: If `ConnectionString` contains a secret (I bet it does), this should be a secret as well?

Same applies to all of the ConnectionStrings in this Routing `Endpoints` object. Possibly we can tell from example ARM templates that use these fields?

File: v2/api/devices/v1api20210702/structure.txt [hunk 8]
```diff
@@ -0,0 +1,841 @@
...
+│   │   │   │           ├── "DeviceJobLifecycleEvents"
+│   │   │   │           ├── "DeviceLifecycleEvents"
+│   │   │   │           ├── "DeviceMessages"
+│   │   │   │           ├── "Invalid"
+│   │   │   │           └── "TwinChangeEvents"
+│   │   │   └── StorageEndpoints: map[string]Object (5 properties)
+│   │   │       ├── AuthenticationType: *Enum (2 values)
+│   │   │       │   ├── "identityBased"
+│   │   │       │   └── "keyBased"
+│   │   │       ├── ConnectionString: *string
```
Comment (matthchr) [inline v2/api/devices/v1api20210702/structure.txt:206]: Secret?

File: v2/api/devices/v1api20210702/structure.txt [hunk 9]
```diff
@@ -0,0 +1,841 @@
...
+│       │   └── UserAssignedIdentities: map[string]Object (2 properties)
+│       │       ├── ClientId: *string
+│       │       └── PrincipalId: *string
+│       ├── Location: *string
+│       ├── Name: *string
+│       ├── Properties: *Object (24 properties)
+│       │   ├── AllowedFqdnList: string[]
+│       │   ├── AuthorizationPolicies: Object (4 properties)[]
+│       │   │   ├── KeyName: *string
+│       │   │   ├── PrimaryKey: *string
```
Comment (matthchr) [inline v2/api/devices/v1api20210702/structure.txt:243]: Should be secret and thus redacted? We might need to talk to them about their Swagger.

File: v2/internal/controllers/crd_devices_iothub_test.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+	devices "github.com/Azure/azure-service-operator/v2/api/devices/v1api20210702"
+
+	"github.com/Azure/azure-service-operator/v2/internal/testcommon"
+	"github.com/Azure/azure-service-operator/v2/internal/util/to"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+func Test_Devices_IotHub_CRUD(t *testing.T) {
+
```
Comment (matthchr) [inline v2/internal/controllers/crd_devices_iothub_test.go:21]: minor: unnecessary newline

File: v2/internal/controllers/crd_devices_iothub_test.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	tc.Expect(retryAfter).To(BeZero())
+	tc.Expect(exists).To(BeFalse())
+
+}
+
+func IotHub_WriteSecrets(tc *testcommon.KubePerTestContext, iotHub *devices.IotHub) {
+	old := iotHub.DeepCopy()
+	iotHubKeysSecret := "iothubkeyssecret"
+	iotHub.Spec.OperatorSpec = &devices.IotHubOperatorSpec{
+
```
Comment (matthchr) [inline v2/internal/controllers/crd_devices_iothub_test.go:72]: minor: Unneeded newline

File: v2/internal/controllers/crd_devices_iothub_test.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+			IotHubOwnerSecondaryKey:       &genruntime.SecretDestination{Name: iotHubKeysSecret, Key: "iotHubOwnerSecondaryKey"},
+			RegistryReadPrimaryKey:        &genruntime.SecretDestination{Name: iotHubKeysSecret, Key: "registryReadPrimaryKey"},
+			RegistryReadSecondaryKey:      &genruntime.SecretDestination{Name: iotHubKeysSecret, Key: "registryReadSecondaryKey"},
+			RegistryReadWritePrimaryKey:   &genruntime.SecretDestination{Name: iotHubKeysSecret, Key: "registryReadWritePrimaryKey"},
+			RegistryReadWriteSecondaryKey: &genruntime.SecretDestination{Name: iotHubKeysSecret, Key: "registryReadWriteSecondaryKey"},
+			ServicePrimaryKey:             &genruntime.SecretDestination{Name: iotHubKeysSecret, Key: "servicePrimaryKey"},
+			ServiceSecondaryKey:           &genruntime.SecretDestination{Name: iotHubKeysSecret, Key: "serviceSecondaryKey"},
+		},
+	}
+	tc.ExportAsSample(iotHub)
```
Comment (matthchr) [inline v2/internal/controllers/crd_devices_iothub_test.go:86]: minor: remove

File: v2/samples/devices/v1api/v1api20210702_iothub.yaml [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        name: iothubkeyssecret
+      registryReadWriteSecondaryKey:
+        key: registryReadWriteSecondaryKey
+        name: iothubkeyssecret
+      servicePrimaryKey:
+        key: servicePrimaryKey
+        name: iothubkeyssecret
+      serviceSecondaryKey:
+        key: serviceSecondaryKey
+        name: iothubkeyssecret
```
Comment (matthchr) [inline v2/samples/devices/v1api/v1api20210702_iothub.yaml:44]: minor: newline at end of file

**Review by matthchr — CHANGES_REQUESTED** [review]

File: v2/api/devices/customizations/iot_hub_extension.go (hunk 2 above)
Comment (super-harsh) [reply v2/api/devices/customizations/iot_hub_extension.go:131]: Link : https://learn.microsoft.com/en-us/rest/api/iothub/iot-hub-resource/list-keys?tabs=HTTP#sharedaccesssignatureauthorizationrule

Dropping a comment in code as well

File: v2/api/devices/v1api20210702/structure.txt (hunk 6 above)
Comment (super-harsh) [reply v2/api/devices/v1api20210702/structure.txt:36]: Thanks for pointing out. Only PrimaryKey and SecondaryKey fields should be marked as read-only and are generated by azure. That said, user can specify their own custom auth policy with the keyName and accessRights and Azure would generate the keys for them. However, the only problem here is, how would we expose those keys to the users once its created by the user?

Comment (theunrepentantgeek) [reply v2/api/devices/v1api20210702/structure.txt:35]: Isn't this what AzureGeneratedSecrets is about? Though, we specify that for the Status of the resource, and it looks like there's a whole array here which is ... fun!

File: v2/api/devices/v1api20210702/structure.txt [hunk 14]
```diff
@@ -0,0 +1,839 @@
...
+│       │   │   ├── "Disabled"
+│       │   │   └── "Enabled"
+│       │   ├── RestrictOutboundNetworkAccess: *bool
+│       │   ├── Routing: *Object (4 properties)
+│       │   │   ├── Endpoints: *Object (4 properties)
+│       │   │   │   ├── EventHubs: Object (9 properties)[]
+│       │   │   │   │   ├── AuthenticationType: *Enum (2 values)
+│       │   │   │   │   │   ├── "identityBased"
+│       │   │   │   │   │   └── "keyBased"
+│       │   │   │   │   ├── ConnectionString: *string
```
[the remark this answers is not in the document]
Comment (matthchr) [reply v2/api/devices/v1api20210702/structure.txt:322]: https://github.com/Azure/azure-service-operator/pull/3034 - let's push this IOTHub resource to 2.2.0 and ship 2.1.0 and we can deal with this and the AKS breaking change then?

[the remark this answers is not in the document]
Comment (super-harsh) [reply v2/api/devices/v1api20210702/structure.txt:322]: I see the same for other connectionStrings as well

**Review by matthchr — APPROVED** [review]

**Pull request merged**